README
NAME
    DOCSIS::ConfigFile - Decodes and encodes DOCSIS config files

VERSION
    0.72

DESCRIPTION
    DOCSIS::ConfigFile is a class which provides functionality to decode and
    encode DOCSIS <http://www.cablelabs.com> (Data over Cable Service
    Interface Specifications) config files.

    This module is used as a layer between any human readable data and the
    binary structure.

    The files are usually served using a TFTP server, after a cable modem
    <http://en.wikipedia.org/wiki/Cable_modem> or MTA (Multimedia Terminal
    Adapter) has recevied an IP address from a DHCP server. These files are
    binary encode using a variety of functions, but all the data in the file
    are constructed by TLVs (type-length-value) blocks. These can be nested
    and concatenated.

    See "CONFIGURATION TREE" in DOCSIS::ConfigFile::Syminfo for full set of
    possible parameters. Create an issue
    <https://github.com/jhthorsen/docsis-configfile/issues> if a parameter
    is missing or invalid.

SYNOPSIS
      use DOCSIS::ConfigFile qw( encode_docsis decode_docsis );

      $data = decode_docsis $bytes;

      $bytes = encode_docsis(
                 {
                   GlobalPrivacyEnable => 1,
                   MaxCPE              => 2,
                   NetworkAccess       => 1,
                   BaselinePrivacy => {
                     AuthTimeout       => 10,
                     ReAuthTimeout     => 10,
                     AuthGraceTime     => 600,
                     OperTimeout       => 1,
                     ReKeyTimeout      => 1,
                     TEKGraceTime      => 600,
                     AuthRejectTimeout => 60,
                     SAMapWaitTimeout  => 1,
                     SAMapMaxRetries   => 4
                   },
                   SnmpMibObject => [
                     {oid => "1.3.6.1.4.1.1.77.1.6.1.1.6.2",    INTEGER => 1},
                     {oid => "1.3.6.1.4.1.1429.77.1.6.1.1.6.2", STRING  => "bootfile.bin"}
                   ],
                   VendorSpecific => {
                     id => "0x0011ee",
                     options => [30 => "0xff", 31 => "0x00", 32 => "0x28"]
                   }
                 }
               );

    See also "CONFIGURATION TREE" in DOCSIS::ConfigFile::Syminfo.

OPTIONAL MODULE
    You can install the SNMP.pm module to translate between SNMP OID
    formats. With the module installed, you can define the "SnmpMibObject"
    like the example below, instead of using numeric OIDs:

      encode_docsis(
        {
          SnmpMibObject => [
            {oid => "docsDevNmAccessIp.1",     IPADDRESS => "10.0.0.1"},
            {oid => "docsDevNmAccessIpMask.1", IPADDRESS => "255.255.255.255"},
          ]
        },
      );

FUNCTIONS
  decode_docsis
      $data = decode_docsis($byte_string);
      $data = decode_docsis(\$path_to_file);

    Used to decode a DOCSIS config file into a data structure. The output
    $data can be used as input to "encode_docsis". Note: $data will only
    contain array-refs if the DOCSIS parameter occur more than once.

  encode_docsis
      $byte_string = encode_docsis(\%data, \%args);

    Used to encode a data structure into a DOCSIS config file. Each of the
    keys in $data can either hold a hash- or array-ref. An array-ref is used
    if the same DOCSIS parameter occur multiple times. These two formats
    will result in the same $byte_string:

      # Only one SnmpMibObject
      encode_docsis({
        SnmpMibObject => { # hash-ref
          oid => "1.3.6.1.4.1.1429.77.1.6.1.1.6.2", STRING => "bootfile.bin"
        }
      })

      # Allow one or more SnmpMibObjects
      encode_docsis({
        SnmpMibObject => [ # array-ref of hashes
          { oid => "1.3.6.1.4.1.1429.77.1.6.1.1.6.2", STRING => "bootfile.bin" }
        ]
      })

    Possible %args:

    *   mta_algorithm

        This argument is required when encoding MTA config files. Can be set
        to either empty string, "sha1" or "md5".

    *   shared_secret

        This argument is optional, but will be used as the shared secret
        used to increase security between the cable modem and CMTS.

COPYRIGHT AND LICENSE
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution DOCSIS-ConfigFile

1.01 Not Released
 - docsis-configfile will detect unknown arguments

1.00 2023-07-22T09:37:59
 - Add support for docsis-configfile --version and -V
 - Add support for piping data in/out from docsis-configfile
 - Improved exceptions by using croak and confess

0.77 2023-07-21T20:15:34
 - Bundled docsisious with this distribution
 - Add /parameters to docsisious
 - Add support for docsis-configfile -c option=value

0.76 2018-10-03T20:01:31+0900
 - Remove DOCSIS::ConfigFile::Syminfo
 - Add missing "ushort_list" encode and decode function
 - Add support for encoding and decoding "SubMgmtFilters" #14

0.75 2018-06-12T11:09:15+0800
 - Add support for "-o" parameter for bundled script

0.74 2018-06-08T11:20:12+0800
 - Bundled script to encode/decode from command line

0.73 2017-08-22T20:55:16+0200
 - Fix invalid test in longdschannellist.t

0.72 2017-08-22T20:47:24+0200
 - Add missing Types for TLV 202 #9
 - Will die when pack()-ing invalid length

0.71 2016-12-11T10:08:41+0100
 - Fix encoding config with IpTos #jhthorsen/app-docsisious#2

0.70 2015-10-15T01:30:14+0200
 - Add support for decode_docsis \$path_to_file
 - Remove deprecated methods

0.69 2015-04-07T21:13:47Z
 - MtaConfigDelimiter is optional when creating MTA config files
 - Add support for eRouter MangemnetServer TLV #5 Contributor: Josh Galvez

0.68 2015-01-05T10:15:21Z
 - Add configuration tree structure to DOCSIS::ConfigFile::Syminfo pod
 - Require Perl 5.10.1

0.67 2015-01-05T02:34:42Z
 - Add test for handling of 0x0A value in VendorSpecific Contributor:
   dragic.dusan@gmail.com, RT#87812

0.66 2015-01-05T02:09:43Z
 - Fix SnmpCpeAccessControl should be uchar instead of uint Contributor:
   Josh Galvez, RT#99468
 - Use Digest::SHA instead of Digest::SHA1 Contributor:
   dragic.dusan@gmail.com, RT#87814

0.65 2015-01-05T02:01:30Z
 - New API: Will deprecate object oriented API with function based
 - Add decode_docsis() and encode_docsis()
 - Add DOCSIS related MIB files
 - Add debug output with DOCSIS_CONFIGFILE_DEBUG
 - Add support for string OID with optional SNMP.pm dependency

0.64 2013-08-16T23:33:13Z
 - Fix handling of 0x0a value in vendorspec()
 - Add support for VendorSpecific TLVs in Service Flow configuration.
   Contributor Dušan Dragić

0.6302 2013-07-25T21:56:00Z
 - Fix RT83013: Using the right value when checking high/low values

0.6301 2013-07-25T21:12:15Z
 - Fix RT87301: "return" binds stronger than "or". Contributor
   rurban@x-ray.at

0.63 2012-10-10T00:49:37Z
 - Fix RT72511 Contributor velladecin@hotmail.com

0.62 2012-04-29T10:56:44Z
 - Add SnmpCpeAccessControl to Syminfo Contributor: Elliot

0.61 2011-10-03T21:58:40Z
 - Fix RT70882: ServiceClassName need to be zero-terminated This is
   optional in input data structure and removed/not visible in decoded tree
   either

0.6004 2010-09-12T21:51:02Z
 - Fix #61267: Encode::int() fails on 64b systems: 1234567890 >> 8 ==
   72057594033105405 != 11954685

0.6003 2010-09-05T20:38:25Z
 - Fix uint() cannot return negative integers Contributor: canni
 - Add int() to handle (possible) signed integers Contributor: canni
 - Fix Decode::string() decodes % as %25
 - Add no_value() dec/encode function
 - Change Decode.pm and Encode.pm to confess on errors
 - Change add_symbol() will confess if the symbol already exists
 - Change Encode functions return a list (wantarray is history)
 - Change examples/docsis.pl is renamed and use JSON instead of YAML. Might
   change to Config::Any in the future.

0.6002 2010-09-04T11:34:58Z
 - Fix dump_symbol_tree() can print large numbers
 - Fix dump_symbol_tree() can also follow "vendorspec"
 - Fix remove duplicates from Syminfo table

0.6001 2010-09-04T04:02:29Z
 - Fix use Carp qw/.../
 - Add support for Syminfo->dump_symbol_tree()
 - Restructured symbol table and documentation
 - Don't need Data::Dumper to run tests

0.60 2010-09-04T00:26:27Z
 - Fix uninitialized warnings in Decode::uint() Contributor: canni
 - Change from logger() and errors() to carp() and confess()
 - Add Encode::mic() to avoid warnings about Undefined decode function for
   PCODE/CODE (0/255)
 - Add DESCRIPTION

0.5901 2010-04-03T13:04:38Z
 - Update t/
 - Update Changes
 - Fix dist structure

0.59 2010-01-23T10:19:00Z
 - Fix symbol table is read from a string

0.58 2009-12-08T22:45:00Z
 - Fix 01-pod-coverage.t

0.57 2009-11-08T16:52:00Z
 - Fix decode of uint

0.56 2009-04-07T15:41:00Z
 - Add decoding of TLV64
 - Fix string(): better detection for hexstrings
 - Add add_symbol() for custom syminfo symbols

0.54 2009-01-08T23:00:00Z
 - Fix SNMP objects starting with "0x30 0x82 ... " fail to decode
 - Update POD and formatting of code

0.52 2008-04-14T12:40:00Z
 - Fix undefined $VERSION.

0.51 2008-04-14T00:19:00Z
 - Update POD

0.50 2008-04-13T21:15:00Z
 - Fix RT34547 should be able to handle ->unlimited values :)
 - Fix RT34548 no warnings when decoding CmMic and CmtsMic
 - Update 10-encode-decode test
 - Change output format: hex-values start with '0x'

0.40 2007-12-05T11:23:00Z
 - Hopefully made a decent release for CPAN.
 - Add example script  which should be quite useful: docsis_yaml.pl.

0.30 2007-12-04T22:32:00Z
 - Update encode/decode test.
 - API change: not compatible with old version.
 - Add support for normal/advanced output. (Sorry for the version jump -
   Really bad typo)

0.02 2007-12-04T16:10:00Z
 - Add deps to makefile.
 - Change error reporting to Log4perl.
 - Add encode/decode test.
 - Move syminfo to DOCSIS::ConfigFile::Syminfo.
 - Update POD.

0.01 2007-12-03T23:49:00Z
 - First release of DOCSIS::ConfigFile



lib/DOCSIS/ConfigFile/Decode.pm
package DOCSIS::ConfigFile::Decode;
use strict;
use warnings;
use bytes;

use Carp 'confess';
use Math::BigInt;
use Socket;

our %SNMP_TYPE = (
  0x02 => ['INTEGER',   \&int],
  0x04 => ['STRING',    \&string],
  0x05 => ['NULLOBJ',   sub { }],
  0x40 => ['IPADDRESS', \&ip],
  0x41 => ['COUNTER',   \&uint],
  0x42 => ['UNSIGNED',  \&uint],
  0x43 => ['TIMETICKS', \&uint],
  0x44 => ['OPAQUE',    \&uint],
  0x46 => ['COUNTER64', \&bigint],
);

sub bigint {
  my @bytes    = unpack 'C*', _test_length(int => $_[0]);
  my $negative = $bytes[0] & 0x80;
  my $int64    = Math::BigInt->new(0);

  # setup int64
  for my $chunk (@bytes) {
    $chunk ^= 0xff if ($negative);
    $int64 = ($int64 << 8) | $chunk;
  }

  if ($negative) {
    $int64 *= -1;
    $int64 -= 1;
  }

  return $int64;
}

sub ether {
  my $length = length $_[0];
  return join '', unpack 'H2' x $length, $_[0] if $length == 6 or $length == 12;
  confess "Invalid ether input. Invalid length ($length)";
}

sub hexstr { '0x' . join '', unpack 'H*', $_[0] }

sub int {
  my @bytes    = unpack 'C*', _test_length(int => $_[0], 'int');
  my $negative = $bytes[0] & 0x80;
  my $int      = 0;

  for my $chunk (@bytes) {
    $chunk ^= 0xff if ($negative);
    $int = ($int << 8) | $chunk;
  }

  if ($negative) {
    $int *= -1;
    $int -= 1;
  }

  return $int;
}

sub ip { inet_ntoa($_[0]) || confess 'inet_ntoa(...) failed to unpack binary string' }
sub mic      {&hexstr}
sub no_value {''}

sub snmp_object {
  my $bin = $_[0];
  my ($byte, $length, $oid, $type, $value);

  # message
  $type = _truncate_and_unpack(\$bin, 'C1');    # 0x30
  $length = _snmp_length(\$bin);

  # oid
  $type   = _truncate_and_unpack(\$bin, 'C1');    # 0x06
  $length = _snmp_length(\$bin);
  $oid    = _snmp_oid(\$bin, $length);

  # value
  $type   = $SNMP_TYPE{_truncate_and_unpack(\$bin, 'C1')};
  $length = _snmp_length(\$bin);
  $value  = $type->[1]->($bin);

  return {oid => $oid, type => $type->[0], value => $value};
}

sub string {

  # not sure why this is able to join - may be removed later
  my $bin = @_ > 1 ? join '', map { chr $_ } @_ : $_[0];

  return hexstr($bin) if $bin =~ /^[^\t\n\r\x20-\xEF]/;
  $bin =~ s/([^\x20-\x24\x26-\x7e])/{ sprintf "%%%02x", ord $1 }/ge;
  return $bin;
}

sub stringz { my $str = string(@_); $str =~ s/%00$//; $str; }
sub uchar { unpack 'C', _test_length(uchar => $_[0], 'char') }

sub uint {
  my @bytes = unpack 'C*', _test_length(uint => $_[0], 'int');
  my $value = 0;

  $value = ($value << 8) | $_ for (@bytes);
  return $value;
}

sub ushort { unpack 'n', _test_length(ushort => $_[0], 'short int') }

sub vendorspec {
  my $bin = $_[0] || '';
  my ($vendor, @ret, $length);

  # extract length (not sure what the first byte is...)
  $bin =~ s/^.(.)//s or confess 'Invalid vendorspec input. Could not extract length';
  $length = unpack 'C', $1;

  # extract vendor
  $bin =~ s/^(.{$length})//s or confess 'Invalid vendorspec input. Could not extract vendor';
  $vendor = sprintf '0x' . ('%02x' x $length), unpack 'C*', $1;

  # extract TLV
  while ($bin =~ s/^(.)(.)//s) {
    my $type   = unpack 'C*', $1;
    my $length = unpack 'C*', $2;

    if ($bin =~ s/^(.{$length})//s) {
      push @ret, {type => $type, length => $length, value => hexstr($1)};
    }
  }

  confess "vendorspec('...') is left with ($length) bytes after decoding" if $length = length $bin;
  return $vendor, \@ret;
}

sub vendor {
  my $bin = shift || '';
  my $length = $bin =~ s/^.(.)//s ? unpack 'C', $1 : 0;
  my ($id, @options);

  if ($bin =~ s/^(.{$length})//s) {
    $id = sprintf "0x@{['%02x' x $length]}", unpack 'C*', $1;
  }

  while ($bin =~ s/^(.)(.)//s) {
    my $type   = unpack 'C*', $1;
    my $length = unpack 'C*', $2;

    $bin =~ s/^(.{$length})//s or next;
    push @options, $type, hexstr($1);
  }

  confess 'Bytes left in vendorspec' if length $bin;
  confess 'Invalid vendorspec' unless defined $id;
  return {id => $id, options => \@options};
}

sub _byte_size {
  return 2  if lc $_[0] eq 'short int';
  return 4  if lc $_[0] eq 'int';
  return 4  if lc $_[0] eq 'long int';
  return 1  if lc $_[0] eq 'char';
  return 4  if lc $_[0] eq 'float';
  return 8  if lc $_[0] eq 'double';
  return 12 if lc $_[0] eq 'long double';
  return 16 if lc $_[0] eq 'md5digest';
}

sub _snmp_length {
  my $length = _truncate_and_unpack($_[0], 'C1');    # length?

  if ($length <= 0x80) {
    return $length;
  }
  elsif ($length == 0x81) {
    return _truncate_and_unpack($_[0], 'C1');
  }
  elsif ($length == 0x82) {
    $length = 0;

    for my $byte (_truncate_and_unpack($_[0], 'C2')) {
      $length = $length << 8 | $byte;
    }

    return $length;
  }

  confess "Too long SNMP length: ($length)";
}

sub _snmp_oid {
  my @bytes = _truncate_and_unpack($_[0], 'C' . $_[1]);
  my @oid   = (0);
  my $subid = 0;

  for my $id (@bytes) {
    if ($subid & 0xfe000000) {
      confess "_snmp_oid(@bytes): Sub-identifier too large: ($subid)";
    }

    $subid = ($subid << 7) | ($id & 0x7f);

    unless ($id & 0x80) {
      confess "_snmp_oid(@bytes): Exceeded max length" if (128 <= @oid);
      push @oid, $subid;
      $subid = 0;
    }
  }

  # the first two sub-id are in the first id
  if ($oid[1] == 0x2b) {    # Handle the most common case
    $oid[0] = 1;
    $oid[1] = 3;
  }
  elsif ($oid[1] < 40) {
    $oid[0] = 0;
  }
  elsif ($oid[1] < 80) {
    $oid[0] = 1;
    $oid[1] -= 40;
  }
  else {
    $oid[0] = 2;
    $oid[1] -= 80;
  }

  return SNMP::translateObj(join '.', @oid) || join '.', @oid if DOCSIS::ConfigFile::CAN_TRANSLATE_OID;
  return join '.', @oid;
}

sub _test_length {
  my $name   = $_[0];
  my $length = length $_[1];

  if (!$length) {
    confess "$name(...) bytestring length is zero";
  }
  if ($_[2]) {
    my $max = _byte_size($_[2]);
    confess "$name(...) bytestring length is invalid: $max < $length" if ($max < $length);
  }

  return $_[1];
}

sub _truncate_and_unpack {
  my ($bin_ref, $type) = @_;
  my $n = ($type =~ /C/ ? 1 : 2) * ($type =~ /(\d+)/)[0];

  if ($$bin_ref =~ s/^(.{$n})//s) {
    return unpack $type, $1;
  }
  else {
    confess "_truncate_and_unpack('...', $type) failed to truncate binary string";
  }
}

1;

=encoding utf8

=head1 NAME

DOCSIS::ConfigFile::Decode - Decode functions for a DOCSIS config-file

=head1 DESCRIPTION

L<DOCSIS::ConfigFile::Decode> has functions which is used to decode binary data
into either plain strings or complex data structures, dependent on the function
called.

=head1 FUNCTIONS

=head2 bigint

Returns a C<Math::BigInt> object.

=head2 ether

Will unpack the input string and return a MAC address in this format:
"00112233" or "00112233445566".

=head2 hexstr

Will unpack the input string and a string with leading "0x", followed
by hexidesimal characters.

=head2 int

Will unpack the input string and return an integer, from -2147483648
to 2147483647.

=head2 ip

Will unpack the input string and return a human readable IPv4 address.

=head2 string

Returns human-readable string, where special characters are "uri encoded".
Example: "%" = "%25" and " " = "%20". It can also return the value from
L</hexstr> if it starts with a weird character, such as C<\x00>.

=head2 stringz

Same as string above. However this string is zero-terminated in encoded
form, but this function remove the last "\0" seen in the string.

=head2 mic

Returns a value, printed as hex.

=head2 no_value

This method will return an empty string. It is used by DOCSIS types, which
has zero length.

=head2 snmp_object

Will take a binary string and decode it into a complex
datastructure, with "oid", "type" and "value".

=head2 uchar

Will unpack the input string and return a short integer, from 0 to 255.

=head2 uint

Will unpack the input string and return an integer, from 0 to 4294967295.

=head2 ushort

Will unpack the input string and return a short integer, from 0 to 65535.

=head2 vendor

Will byte-encode a complex vendorspec datastructure.

=head2 vendorspec

Will unpack the input string and return a complex datastructure,
representing the vendor specific data.

=head1 SEE ALSO

L<DOCSIS::ConfigFile>

=cut
